dawn/src/dawn/IIR/ASTStmt.cpp

```diff
@@ -13,16 +13,89 @@
+//     IIRStmtData
+//===------------------------------------------------------------------------------------------===//
+
+IIRStmtData::IIRStmtData(const IIRStmtData& other) {
+  StackTrace = other.StackTrace ? std::make_optional(*other.StackTrace) : other.StackTrace;
+  CallerAccesses =
+      other.CallerAccesses ? std::make_optional(*other.CallerAccesses) : other.CallerAccesses;
+  CalleeAccesses =
+      other.CalleeAccesses ? std::make_optional(*other.CalleeAccesses) : other.CalleeAccesses;
+}
+
+bool IIRStmtData::operator==(const IIRStmtData& rhs) {
+  return StackTrace == rhs.StackTrace && CallerAccesses == rhs.CallerAccesses &&
+         CalleeAccesses == rhs.CalleeAccesses;
+}
 bool IIRStmtData::operator!=(const IIRStmtData& rhs) { return !(*this == rhs); }
 
 std::unique_ptr<ast::StmtData> IIRStmtData::clone() const {
   return std::make_unique<IIRStmtData>(*this);
 }
+
+//===------------------------------------------------------------------------------------------===//
+//     VarDeclStmtData
+//===------------------------------------------------------------------------------------------===//
+
+VarDeclStmtData::VarDeclStmtData(const VarDeclStmtData& other) : IIRStmtData(other) {
+  AccessID = other.AccessID ? std::make_optional(*other.AccessID) : other.AccessID;
+}
+
+bool VarDeclStmtData::operator==(const VarDeclStmtData& rhs) {
+  return IIRStmtData::operator==(rhs) && AccessID == rhs.AccessID;
+}
+bool VarDeclStmtData::operator!=(const VarDeclStmtData& rhs) { return !(*this == rhs); }
+
+std::unique_ptr<ast::StmtData> VarDeclStmtData::clone() const {
+  return std::make_unique<VarDeclStmtData>(*this);
+}
+
+//===------------------------------------------------------------------------------------------===//
+//     computeMaximumExtents
+//===------------------------------------------------------------------------------------------===//
+
+std::optional<Extents> computeMaximumExtents(Stmt& stmt, const int accessID) {
+  std::optional<Extents> extents;
+
+  const auto& callerAccesses = stmt.getData<IIRStmtData>().CallerAccesses;
+
+  if(callerAccesses->hasReadAccess(accessID) || callerAccesses->hasWriteAccess(accessID)) {
+    extents = std::optional<Extents>();
+
+    if(callerAccesses->hasReadAccess(accessID)) {
+      if(!extents)
+        extents = std::make_optional(callerAccesses->getReadAccess(accessID));
+      else
+        extents->merge(callerAccesses->getReadAccess(accessID));
+    }
+    if(callerAccesses->hasWriteAccess(accessID)) {
+      if(!extents)
+        extents = std::make_optional(callerAccesses->getWriteAccess(accessID));
+      else
+        extents->merge(callerAccesses->getWriteAccess(accessID));
+    }
+  }
+
+  for(auto const& child : stmt.getChildren()) {
+    auto childExtent = computeMaximumExtents(*child, accessID);
+    if(!childExtent)
+      continue;
+    if(extents)
+      extents->merge(*childExtent);
+    else
+      extents = childExtent;
```

Comment:
`merge` returns void, so can't assign to `extents` ... but it can become
```
if(auto childExtent = computeMaximumExtents(*child, accessID))
      extents ? extents->merge(*childExtent) : extents = childExtent;
```

Reply:
Tried it, but the compiler doesn't like it. Need if-else.

Reply:
Yes because types are not compatible; and I also wouldn't do it. In case you would want to do it (what you really shouldn't):
```
extents = extents ? (extents->merge(*childExtent), extents) : childExtent;
```
This should work.

But I think as long as merge is inplace, we need to have an if.  I would still put the `auto childExtent =` into the if, because then we don't need the `continue` which usually makes the code-flow harder to understand.